src/global/ValueId.h

```diff
@@ -156,61 +157,54 @@ class ValueId {
-    auto compareVocabAndLocalVocab =
-        [](IdProxy vocabIndex,
-           ::LocalVocabIndex localVocabIndex) -> std::strong_ordering {
-      auto [lowerBound, upperBound] = localVocabIndex->positionInVocab();
-      if (vocabIndex < lowerBound) {
-        return std::strong_ordering::less;
-      } else if (vocabIndex >= upperBound) {
-        return std::strong_ordering::greater;
-      } else {
-        return std::strong_ordering::equal;
-      }
-    };
-    // GCC 11 issues a false positive warning here, so we try to avoid it by
-    // being over-explicit about the branches here.
-    if ((type == VocabIndex || type == EncodedVal) &&
-        otherType == LocalVocabIndex) {
-      return compareVocabAndLocalVocab(IdProxy::make(getBits()),
-                                       other.getLocalVocabIndex());
-    } else if (type == LocalVocabIndex &&
-               (otherType == VocabIndex || otherType == EncodedVal)) {
-      auto inverseOrder = compareVocabAndLocalVocab(
-          IdProxy::make(other.getBits()), getLocalVocabIndex());
-      return 0 <=> inverseOrder;
-    }
-
-    // One of the types is `LocalVocab`, and the other one is a non-string type
-    // like `Integer` or `Undefined. Then the comparison by bits automatically
-    // compares by the datatype.
-    return _bits <=> other._bits;
-  }
+  QL_DEFINE_CUSTOM_THREEWAY_OPERATOR_LOCAL_CONSTEXPR(
+      ValueId, (const ValueId& other) const, {
+        using enum Datatype;
+        auto type = getDatatype();
+        auto otherType = other.getDatatype();
+        if (type != LocalVocabIndex && otherType != LocalVocabIndex) {
+          return ql::compareThreeWay(_bits, other._bits);
+        }
+        if (type == LocalVocabIndex && otherType == LocalVocabIndex)
+            [[unlikely]] {
+          return ql::compareThreeWay(*getLocalVocabIndex(),
+                                     *other.getLocalVocabIndex());
+        }
+
+        // GCC 11 issues a false positive warning here, so we try to avoid it by
+        // being over-explicit about the branches here.
+        if ((type == VocabIndex || type == EncodedVal) &&
+            otherType == LocalVocabIndex) {
+          return compareVocabAndLocalVocab(
+              LocalVocabEntry::IdProxy::make(getBits()),
+              other.getLocalVocabIndex());
+        } else if (type == LocalVocabIndex &&
+                   (otherType == VocabIndex || otherType == EncodedVal)) {
+          auto inverseOrder = compareVocabAndLocalVocab(
+              LocalVocabEntry::IdProxy::make(other.getBits()),
+              getLocalVocabIndex());
+
+          return ql::compareThreeWay(0, inverseOrder);
+        }
+
+        // One of the types is `LocalVocab`, and the other one is a non-string
+        // type like `Integer` or `Undefined. Then the comparison by bits
+        // automatically compares by the datatype.
+        return ql::compareThreeWay(_bits, other._bits);
+      })
```

Comment:
The macros that have a full function body as an argument lead to really really strange formatting. For those manual cases I would suggest that you spell out the `compareThreeway` member function manually, and then have a macro that "just" assumes the presence of this function and synthesizes the operators.

```
auto compareThreeway(const ValueId& other) { return ql::compareThreeWay(this->something, other.something();
}
QL_DEFINE_CUSTOM_THREEWAY_CONSTEXPR_OR_NOT(ValueId);
```

That way we have the shortest syntax in all cases:
Either 
1. Just list all the members, or
2. Do custom logic which was there anyway (in a readable way) + a single macro line.